// angular-pipes/esm2015/boolean/conditions.pipe.js
import * as tslib_1 from "tslib";
import { Pipe } from '@angular/core';
let IsGreaterPipe = class IsGreaterPipe {
    transform(first, second) {
        return first > second;
    }
};
IsGreaterPipe = tslib_1.__decorate([
    Pipe({
        name: 'greater'
    })
], IsGreaterPipe);
export { IsGreaterPipe };
let IsGreaterOrEqualPipe = class IsGreaterOrEqualPipe {
    transform(first, second) {
        return first >= second;
    }
};
IsGreaterOrEqualPipe = tslib_1.__decorate([
    Pipe({
        name: 'greaterOrEqual'
    })
], IsGreaterOrEqualPipe);
export { IsGreaterOrEqualPipe };
let IsLessPipe = class IsLessPipe {
    transform(first, second) {
        return first < second;
    }
};
IsLessPipe = tslib_1.__decorate([
    Pipe({
        name: 'less'
    })
], IsLessPipe);
export { IsLessPipe };
let IsLessOrEqualPipe = class IsLessOrEqualPipe {
    transform(first, second) {
        return first <= second;
    }
};
IsLessOrEqualPipe = tslib_1.__decorate([
    Pipe({
        name: 'lessOrEqual'
    })
], IsLessOrEqualPipe);
export { IsLessOrEqualPipe };
let IsEqualPipe = class IsEqualPipe {
    transform(first, second) {
        return first == second;
    }
};
IsEqualPipe = tslib_1.__decorate([
    Pipe({
        name: 'equal'
    })
], IsEqualPipe);
export { IsEqualPipe };
let IsNotEqualPipe = class IsNotEqualPipe {
    transform(first, second) {
        return first != second;
    }
};
IsNotEqualPipe = tslib_1.__decorate([
    Pipe({
        name: 'notEqual'
    })
], IsNotEqualPipe);
export { IsNotEqualPipe };
let IsIdenticalPipe = class IsIdenticalPipe {
    transform(first, second) {
        return first === second;
    }
};
IsIdenticalPipe = tslib_1.__decorate([
    Pipe({
        name: 'identical'
    })
], IsIdenticalPipe);
export { IsIdenticalPipe };
let IsNotIdenticalPipe = class IsNotIdenticalPipe {
    transform(first, second) {
        return first !== second;
    }
};
IsNotIdenticalPipe = tslib_1.__decorate([
    Pipe({
        name: 'notIdentical'
    })
], IsNotIdenticalPipe);
export { IsNotIdenticalPipe };
