// angular-pipes/esm2015/boolean/types.pipe.js
import * as tslib_1 from "tslib";
import { Pipe } from '@angular/core';
import { isNull, isNil, isUndefined, isFunction, isArray, isString, isObject, isNumber } from '../utils/utils';
let IsNullPipe = class IsNullPipe {
    transform(value) {
        return isNull(value);
    }
};
IsNullPipe = tslib_1.__decorate([
    Pipe({
        name: 'isNull'
    })
], IsNullPipe);
export { IsNullPipe };
let IsUndefinedPipe = class IsUndefinedPipe {
    transform(value) {
        return isUndefined(value);
    }
};
IsUndefinedPipe = tslib_1.__decorate([
    Pipe({
        name: 'isUndefined'
    })
], IsUndefinedPipe);
export { IsUndefinedPipe };
let IsNilPipe = class IsNilPipe {
    transform(value) {
        return isNil(value);
    }
};
IsNilPipe = tslib_1.__decorate([
    Pipe({
        name: 'isNil'
    })
], IsNilPipe);
export { IsNilPipe };
let IsFunctionPipe = class IsFunctionPipe {
    transform(value) {
        return isFunction(value);
    }
};
IsFunctionPipe = tslib_1.__decorate([
    Pipe({
        name: 'isFunction'
    })
], IsFunctionPipe);
export { IsFunctionPipe };
let IsNumberPipe = class IsNumberPipe {
    transform(value) {
        return isNumber(value);
    }
};
IsNumberPipe = tslib_1.__decorate([
    Pipe({
        name: 'isNumber'
    })
], IsNumberPipe);
export { IsNumberPipe };
let IsStringPipe = class IsStringPipe {
    transform(value) {
        return isString(value);
    }
};
IsStringPipe = tslib_1.__decorate([
    Pipe({
        name: 'isString'
    })
], IsStringPipe);
export { IsStringPipe };
let IsArrayPipe = class IsArrayPipe {
    transform(value) {
        return isArray(value);
    }
};
IsArrayPipe = tslib_1.__decorate([
    Pipe({
        name: 'isArray'
    })
], IsArrayPipe);
export { IsArrayPipe };
let IsObjectPipe = class IsObjectPipe {
    transform(value) {
        return isObject(value);
    }
};
IsObjectPipe = tslib_1.__decorate([
    Pipe({
        name: 'isObject'
    })
], IsObjectPipe);
export { IsObjectPipe };
let IsDefinedPipe = class IsDefinedPipe {
    transform(value) {
        return !isUndefined(value);
    }
};
IsDefinedPipe = tslib_1.__decorate([
    Pipe({
        name: 'isDefined'
    })
], IsDefinedPipe);
export { IsDefinedPipe };
