// angular-pipes/esm5/boolean/types.pipe.js
import * as tslib_1 from "tslib";
import { Pipe } from '@angular/core';
import { isNull, isNil, isUndefined, isFunction, isArray, isString, isObject, isNumber } from '../utils/utils';
var IsNullPipe = /** @class */ (function () {
    function IsNullPipe() {
    }
    IsNullPipe.prototype.transform = function (value) {
        return isNull(value);
    };
    IsNullPipe = tslib_1.__decorate([
        Pipe({
            name: 'isNull'
        })
    ], IsNullPipe);
    return IsNullPipe;
}());
export { IsNullPipe };
var IsUndefinedPipe = /** @class */ (function () {
    function IsUndefinedPipe() {
    }
    IsUndefinedPipe.prototype.transform = function (value) {
        return isUndefined(value);
    };
    IsUndefinedPipe = tslib_1.__decorate([
        Pipe({
            name: 'isUndefined'
        })
    ], IsUndefinedPipe);
    return IsUndefinedPipe;
}());
export { IsUndefinedPipe };
var IsNilPipe = /** @class */ (function () {
    function IsNilPipe() {
    }
    IsNilPipe.prototype.transform = function (value) {
        return isNil(value);
    };
    IsNilPipe = tslib_1.__decorate([
        Pipe({
            name: 'isNil'
        })
    ], IsNilPipe);
    return IsNilPipe;
}());
export { IsNilPipe };
var IsFunctionPipe = /** @class */ (function () {
    function IsFunctionPipe() {
    }
    IsFunctionPipe.prototype.transform = function (value) {
        return isFunction(value);
    };
    IsFunctionPipe = tslib_1.__decorate([
        Pipe({
            name: 'isFunction'
        })
    ], IsFunctionPipe);
    return IsFunctionPipe;
}());
export { IsFunctionPipe };
var IsNumberPipe = /** @class */ (function () {
    function IsNumberPipe() {
    }
    IsNumberPipe.prototype.transform = function (value) {
        return isNumber(value);
    };
    IsNumberPipe = tslib_1.__decorate([
        Pipe({
            name: 'isNumber'
        })
    ], IsNumberPipe);
    return IsNumberPipe;
}());
export { IsNumberPipe };
var IsStringPipe = /** @class */ (function () {
    function IsStringPipe() {
    }
    IsStringPipe.prototype.transform = function (value) {
        return isString(value);
    };
    IsStringPipe = tslib_1.__decorate([
        Pipe({
            name: 'isString'
        })
    ], IsStringPipe);
    return IsStringPipe;
}());
export { IsStringPipe };
var IsArrayPipe = /** @class */ (function () {
    function IsArrayPipe() {
    }
    IsArrayPipe.prototype.transform = function (value) {
        return isArray(value);
    };
    IsArrayPipe = tslib_1.__decorate([
        Pipe({
            name: 'isArray'
        })
    ], IsArrayPipe);
    return IsArrayPipe;
}());
export { IsArrayPipe };
var IsObjectPipe = /** @class */ (function () {
    function IsObjectPipe() {
    }
    IsObjectPipe.prototype.transform = function (value) {
        return isObject(value);
    };
    IsObjectPipe = tslib_1.__decorate([
        Pipe({
            name: 'isObject'
        })
    ], IsObjectPipe);
    return IsObjectPipe;
}());
export { IsObjectPipe };
var IsDefinedPipe = /** @class */ (function () {
    function IsDefinedPipe() {
    }
    IsDefinedPipe.prototype.transform = function (value) {
        return !isUndefined(value);
    };
    IsDefinedPipe = tslib_1.__decorate([
        Pipe({
            name: 'isDefined'
        })
    ], IsDefinedPipe);
    return IsDefinedPipe;
}());
export { IsDefinedPipe };
